File: android/src/main/java/com/flutter_google_cast_plugin/CastStateStreamHandler.java
package com.flutter_google_cast_plugin;

import android.util.Log;

import com.google.android.gms.cast.framework.CastContext;
import com.google.android.gms.cast.framework.CastState;
import com.google.android.gms.cast.framework.CastStateListener;

import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.PluginRegistry;

public class CastStateStreamHandler implements EventChannel.StreamHandler {
    private static final String TAG = "CastState";

    public static void registerWith(PluginRegistry.Registrar registrar) {
        final EventChannel channel = new EventChannel(registrar.messenger(), "flutter_google_cast_plugin.CastState");
        channel.setStreamHandler(new CastStateStreamHandler());
        Log.i(TAG, "registered plugin");
    }

    @Override
    public void onListen(Object o, EventChannel.EventSink sink) {
        Log.d(TAG, "onListen");

        // store sink reference
        eventSink = sink;

        // listen for cast state updates
        CastContext context = CastContext.getSharedInstance();
        context.addCastStateListener(castStateListener);

        // synchonize last state
        lastState = context.getCastState();
        sink.success(lastState);
    }

    @Override
    public void onCancel(Object o) {
        Log.d(TAG, "onCancel");

        // stop listening
        CastContext context = CastContext.getSharedInstance();
        context.removeCastStateListener(castStateListener);

        // remove sink reference
        eventSink = null;
    }

    private int lastState = CastState.NO_DEVICES_AVAILABLE;
    private EventChannel.EventSink eventSink;

    private CastStateListener castStateListener = new CastStateListener() {
        @Override
        public void onCastStateChanged(int newState) {
            Log.v(TAG, "castStateListener newState="+newState);

            lastState = newState;
            if (eventSink != null)
                eventSink.success(newState);
        }
    };
}

File: ios/Classes/CastStateStreamHandler.swift
import Flutter
import GoogleCast

public class CastStateStreamHandler: NSObject, FlutterStreamHandler {
    
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterEventChannel(name: "flutter_google_cast_plugin.CastState", binaryMessenger: registrar.messenger())
    channel.setStreamHandler(CastStateStreamHandler())
  }

  private var lastState = GCKCastState.noDevicesAvailable;
  private var eventSink: FlutterEventSink?

  public func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
    print("CastState onListen")
    
    // store sink reference
    self.eventSink = eventSink
    
    // listen for cast state updates
    let context = GCKCastContext.sharedInstance()
    NotificationCenter.default.addObserver(self, selector: #selector(onCastStateChanged), name: NSNotification.Name.gckCastStateDidChange, object: GCKCastContext.sharedInstance());
    
    // synchronize last state
    self.lastState = context.castState
    eventSink(self.lastState.rawValue)
    
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    print("CastState onCancel")
    
    // stop listening
    NotificationCenter.default.removeObserver(self, name: NSNotification.Name.gckCastStateDidChange, object: GCKCastContext.sharedInstance())
    
    // remove sink reference
    self.eventSink = nil
    
    return nil
  }
    
    @objc func onCastStateChanged(_: Notification) {
        print("castStateListener newState=\(String(describing:GCKCastContext.sharedInstance().castState.rawValue)) eventSink=\(String(describing:eventSink))")
        
        self.lastState = GCKCastContext.sharedInstance().castState
        eventSink?(self.lastState.rawValue + 1)
    }
}
